packages/vscode-extension/src/plugins/redux-devtools-plugin/redux-devtools-plugin.ts

```diff
@@ -1,91 +1,84 @@
+    window.registerWebviewViewProvider(
+      `${REDUX_PLUGIN_PREFIX}.view`,
+      new ReduxDevToolsPluginWebviewProvider(extensionContext),
+      { webviewOptions: { retainContextWhenHidden: true } }
+    )
   );
-
-  return webviewProvider;
 }
 
-export const createReduxDevtools = (toolsManager: ToolsManager): ToolPlugin => {
-  const webViewProvider = initializeReduxDevPlugin();
-  let disposables: Disposable[] = [];
-  let disposed = false;
+export class ReduxDevtoolsPlugin implements ToolPlugin {
+  public readonly id: ToolKey = REDUX_PLUGIN_ID;
+  public readonly label = "Redux DevTools";
 
-  let proxyDevtoolsListener: null | ((event: string, payload: any) => void) = null;
-  webViewProvider?.setListener((webview) => {
-    if (disposed) {
-      return;
-    }
-    proxyDevtoolsListener = (event: string, payload: any) => {
-      if (event === REDUX_PLUGIN_ID) {
-        webview.webview.postMessage({
-          scope: event,
-          data: payload,
-        });
-      }
-    };
+  public available = false;
+  public readonly persist = true;
 
-    toolsManager.devtools.addListener(proxyDevtoolsListener);
+  private connectedWebview?: Webview;
 
-    disposables.push(
-      webview.webview.onDidReceiveMessage((message) => {
-        const { scope, ...data } = message;
-        toolsManager.devtools.send(scope, data);
-      })
-    );
-    disposables.push(
-      webview.onDidChangeVisibility(() =>
-        reportToolVisibilityChanged(REDUX_PLUGIN_ID, webview.visible)
-      )
-    );
-  });
+  constructor(private readonly devtools: Devtools) {
+    initialize();
+  }
 
-  function dispose() {
-    if (!disposed) {
-      if (proxyDevtoolsListener) {
-        toolsManager.devtools.removeListener(proxyDevtoolsListener);
-      }
-      disposeAll(disposables);
-      disposed = true;
+  devtoolsListener = (event: string, payload: any) => {
+    if (event === REDUX_PLUGIN_ID) {
+      this.connectedWebview?.postMessage({
+        scope: event,
+        data: payload,
+      });
+    } else if (event === "RNIDE_appReady" && this.connectedWebview) {
+      // Sometimes, the messaging channel (devtools) is established only after
+      // the Redux store is created and after it sends the first message. In that
+      // case, the "start" event never makes it to the webview.
+      // To workaround this, we use "appReady" event which is sent after the messaging
+      // channel is established. We then force reload the webview with redux devtools
+      // which causes the devtools to initialize a new session and, as a consequence force the store
+      // to reconnect.
+      const html = this.connectedWebview.html;
+      this.connectedWebview.html = "";
+      this.connectedWebview.html = html;
     }
+  };
+
+  connectDevtoolsWebview(webview: Webview) {
+    this.connectedWebview = webview;
+    webview.onDidReceiveMessage((message) => {
+      const { scope, ...data } = message;
+      this.devtools.send(scope, data);
+    });
   }
```

Comment:
What's wrong with getting the `webview` through `webViewProvider.setListener`?

Reply:
I don't think there's anything inherently wrong. The previous approach had some timing related issue when it comes to that and expected certain callbacks to run in particular order. So I'd assume this was fixable but since we already had a very similar pattern implemented in network that I knew worked, I decided to follow the same approach.